lexer.go

```
@@ -382,7 +434,10 @@ func (l *Lexer) Ignore() {
 // The caller can request that the lexer stop by cancelling ctx.
 func (l *Lexer) Lex(ctx context.Context) error {
 	// Set the channel to support calls back into the Lexer.
-	defer close(l.tokens)
+	defer func() {
+		l.tokens <- &TokenEOF
+		close(l.tokens)
+	}()
 
 	for l.state != nil {
```

Should `Lex` check for `l.err` every iteration and fail early if it is non-nil?